web/src/message_fetch.js

```diff
@@ -452,8 +452,11 @@ export function initialize(home_view_loaded) {
         if (message_lists.home.selected_id() === -1 && !message_lists.home.visibly_empty()) {
             // We fall back to the closest selected id, as the user
             // may have removed a stream from the home view while we
-            // were loading data.
+            // were loading data. We do not mark as read because the
+            // user may not have the home view as their current view
+            // (for reloads) or their default view (for initial load).
             message_lists.home.select_id(data.anchor, {
+                mark_read: false,
```

Comment:
Ooh, I suspect this change explains and fixes this longstanding bug: #19329 / #16986.

Reply:
So, a potential reproducer for those issues on `main`:
- Mark x number of messages unread for Cordelia (can me a mix of direct messages and stream messages)
- Open a tab for her in **All messages** view (doesn't have to be the default view) and pin it (I used Firefox in Linux). Make sure that the pinned tab is such that messages can be marked as read.
- Close Firefox
- Open Firefox

I see the pinned tab with the initial unread count I expect, and then after a few seconds of the tab reloading, it will go down by one. Note that I have no other browsers/tabs open with Zulip Dev.

And when on the branch with these changes, the above does not happen. The unread count won't go down until I click on the tab. It might be the fix here and it might be the fix with the window.focus in the second commit.

Reply:
Mmmm... that reproduces on `main` in a stream view with unread messages, so it's a bug but probably not fixed by this change.

Reply:
Cool, that's a great reproducer -- I guess in particular what's happening there is the Firefox tab pinning behavior is a way that you have a noninteractive reload via a different code path.

And now when I read the logic at the top of `activity.js`, at least the logic there is notably not considering this possibility. That logic is I think not actually involved in marking things as read, but suggests we might have a broader bug here.

```
// When you open Zulip in a new browser window, client_is_active                                        
// should be true.  When a server-initiated reload happens, however,                                    
// it should be initialized to false.  We handle this with a check for                                  
// whether the window is focused at initialization time.                                                
export let client_is_active = document.hasFocus && document.hasFocus();         

// new_user_input is a more strict version of client_is_active used                                     
// primarily for analytics.  We initialize this to true, to count new                                   
// page loads, but set it to false in the onload function in reload.js                                  
// if this was a server-initiated-reload to avoid counting a                                            
// server-initiated reload as user activity.                                                            
export let new_user_input = true;                                                                       
```

Let's definitely dig deeper into how to fix that reproducer -- I think it's one of our more important "we've never been able to figure out why this happens" bugs, so fixing it would be amazing. 

@zbenjamin do you by any chance user Firefox? I think this might explain the unread eating behavior you mentioned a few months back that we couldn't figure out.

Reply:
So, we've got #26299 up for a page reload marking first message as read, which is a variation on this reproducer. 

Here's the conversation about that issue on CZO: https://chat.zulip.org/#narrow/stream/6-frontend/topic/messages.20marked.20as.20read.20in.20open.20tabs.20when.20page.20reloaded/near/1611162.

I think that it's related to the refactoring we did in #25387, specifically commit 8f1c3a0fa1e81e5790460e39ab261c8016cea0d2. It would be good to get your thoughts @timabbott in that conversation/issue.